File: CVRP.py
######################
######## CVRP ########
######################

import pandas as pd
import math as m
import random


######################
##### Read Files #####
######################

capacity = 12
l = [(id,x,y,demand)]

######################
### Aux  Functions ###
######################

def d(a,b):
	return m.sqrt((a[1] - a[2])**2 + (b[1] - b[2])**2)

def sumDemand(l):
	s = 0
	for i in l:
		s += i[3]

def closestNode(node,l,capacity):
	dist_min = 0
	node_min = 0
	for i in l:
		if i[0] != node[0]:
			dist_i = d(node,l)
			if dist_i < dist_min and i[3] > 0:
				dist_min = dist_i
				node_min = i
	return node_min

def genSolution(l,total_demand,capacity):
	solution = [l[0]]
	c = sumDemand(l)
	capacity_i = capacity 
	while c > 0:
		node_i = solution[-1]
		next_node = closestNode(node_i,l,capacity_i)
		solution.append(next_node)
		if next_node[3] >= capacity_i:
			l[next_node[0]][3] -= capacity_i
			capacity_i = 0
			solution.append(l[0])
		else:
			capacity_i -= next_node[3]
			l[next_node[0]][3] = 0
	return solution

def totalCost(solution):
	cost = 0
	for i in range(len(solution)-1):
		cost += d(solution[i+1],solution[i])
	return cost

def deltaCost(solution):
	cost = []
	for i in range(len(solution)-1):
		cost.append((d(solution[i+1],solution[i]),i+1))
	return cost.sort(key=lambda x:x[0])

def getDepots(solution):
	l = []
	c = 0
	for i in solution:
		if i[0] == 0:
			l.append(c)
		c += 1


######################
# Destroy Heuristics #
######################

destroy_heuistics = []

def randomRemoval(solution,alpha):
	r = random.randint(int((len(solution)-1)*alpha))
	for i in r:
		index = random.randint(1,len(solution) - 1)
		solution[index] = ()
	return solution

def worstRemoval(solution,alpha):
	r = random.randint(int((len(solution)-1)*alpha))
	cost = deltaCost(solution)[-1:-r]
	for i in range(len(solution)):
		if solution[i] in cost:
			solution[i] = ()
	return solution

def relatedRemoval(solution):
	depots = getDepots(solution)
	r = random.randint(len(depots)-2)
	for i in range(depots[r],depots[r+1]):
		solution[i] = ()
	return solution


######################
#  Repair Heuristics #
######################

repair_heuristics = []

def basicGreedy(solution):
	return


######################
######## ALNS ########
######################

def 

File: CVRP.cpp
#include <iostream>
#include <cmath>
using namespace std;


struct Node {
	int id;
	int x;
	int y;
	int demand;
};

int capacity;
int n = 100;
struct Node nodes[100];


/* Destroy Heuristics */



/* Repair Heuristics */



/* ALNS */


/* AUX Functions */

float d(struct Node a, struct Node b) {
	float distance;
	distance = sqrt(pow((a.x - b.x),2) + pow((a.y - b.y),2));
	return distance;
};

int closestPossibleNode(struct Node node, struct Node array[], int capacity) {
	int n = node.id;
	float dist_i;
	float dist_min = pow(10,5);
	int n_i;
	int demand_i;
	int node_min = 0;
	int i = 0;
	for (int i = 0; i < n; i++) {
		n_i = array[i].id;
		demand_i = array[i].demand;
		if (n_i != n && demand_i < capacity) {
			dist_i = d(node, array[i]);
			if (dist_i < dist_min) {
				dist_min = dist_i;
				node_min = array[i].id;
			}
		}
	}
	return node_min;
};

int totalDemand(struct Node array[]) {
	int d = 0;
	for (int i = 0; i < n; i++) {
		d += array[i].demand;
	}
};

void initialSolution() {
	struct Node *routes[n];
	int total_demand = totalDemand(nodes);
	int c = 0;
	while (c < total_demand) { 
		struct Node node_i = nodes[0];
		int capacity_i = capacity;
		struct Node route_i[n]; 
	}
};


float routeCost(struct Node array[]) {
	float cost = 0;
	int id;
	for (int i = 0; i < n - 1; i++) {
		id = array[i+1].id;
		if (id == 3141592) {
			break;
		}
		cost += d(array[i+1], array[i]);
	}
	return cost;
};


bool lastItem(struct Node array[]) {
	bool l;
	if (array[0].id == 3141592) {
		l = true;
	}
	else {
		l = false;
	}
	return l;
}

float totalCost(struct Node *array[]) {
	float cost = 0;
	bool l;
	for (int i = 0; i < n ; i++) {
		l = lastItem(*(array+i));
		if (l == true) {
			break;
		}
		else {
			cost += routeCost(*(array+i));
		}
	}
	return cost;
};

int closestPossibleRoute(struct Node node, struct Node *array[]) {
	float dist = pow(10,5);
	int closest_route = 0;
	float d_i;
	int demand = node.demand;
	for (int i = 0; i < n; i++) {
		if (routeCost(*(array+i)) + demand <= capacity) {
			for (int j = 0; j < n; j++) {
				d_i = d(node, *(array+i)[j]);
				if (d_i < dist) {
					dist = d_i;
					closest_route = i;
				}
			}
		}
	}
	return closest_route;
};




struct Node *l[3];
int main()
{
	struct Node a[3] = {{0,10,15,25},{1,11,18,29},{3141592,0,0,0}};
	struct Node b[3] = {{0,10,15,25},{3,10,15,29},{3141592,0,0,0}};
	struct Node c[1] = {{3141592,0,0,0}};
	l[0] = a;
	l[1] = b;
	l[2] = c;
	float cost = totalCost(l);
	cout << cost << endl;
}